### PR TITLE
Use lazy loading

### LINKED ISSUE #20: Use lazy loading
In `Book.java` for the `authors` field, change the fetch type to lazy

### PR DESCRIPTION
**Pull request by chetanchilhate**
## Summary of change

Lazy load authors in book and books in author

## Related issue

Closes #20 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-jcijyenp-JiKFGBv62FIPoFnvOW6Ubg)

### CODE REVIEW

**Comment by chetanchilhate** [comment]
@knjk04 let me if any change is needed

File: src/main/java/com/karankumar/booksapi/repository/BookRepository.java [hunk 1]
```diff
@@ -16,7 +16,13 @@
...
 
 import com.karankumar.booksapi.model.Book;
+import java.util.List;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.CrudRepository;
 
 public interface BookRepository extends CrudRepository<Book, Long> {
+  
+    @Query("SELECT DISTINCT b FROM Book b LEFT JOIN FETCH b.authors")
+    List<Book> findAllBooks();
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/repository/BookRepository.java:26]: Could you please also write a test for this to check whether this works as intended?

File: src/main/java/com/karankumar/booksapi/repository/AuthorRepository.java [hunk 2]
```diff
@@ -16,7 +16,13 @@
...
 
 import com.karankumar.booksapi.model.Author;
+import java.util.List;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.CrudRepository;
 
 public interface AuthorRepository extends CrudRepository<Author, Long> {
+  
+    @Query("SELECT DISTINCT a FROM Author a LEFT JOIN FETCH a.books")
+    List<Author> findAllAuthors();
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/repository/AuthorRepository.java:26]: Could you please also write a test for this to check whether this works as intended?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this! Could you please write tests for the JPQL queries you wrote?

File: src/main/java/com/karankumar/booksapi/repository/BookRepository.java (hunk 1 above)
Comment (chetanchilhate) [reply src/main/java/com/karankumar/booksapi/repository/BookRepository.java:26]: integration test or unit test

Comment (knjk04) [reply src/main/java/com/karankumar/booksapi/repository/BookRepository.java:26]: Unit test, please. So populate the repository and then check whether it find all of the books

**Comment by chetanchilhate** [comment]
@knjk04  added test cases have look 
while writing test cases, I found one issue with Book constructor we are checking not null we should also check list should not be empty if does we should throw IllegalArgumentException book must have at least one author

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+@ExtendWith(SpringExtension.class)
+@DataJpaTest
+public class AuthorRepositoryTest {
+  
+    @Autowired
+    private BookRepository bookRepository;
+    
+    @Autowired
+    private AuthorRepository authorRepository;
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:42]: Could you please change this to use constructor injection?

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import com.karankumar.booksapi.model.Author;
+import com.karankumar.booksapi.model.Book;
+import com.karankumar.booksapi.model.BookFormat;
+import com.karankumar.booksapi.model.BookGenre;
+import com.karankumar.booksapi.model.Publisher;
+
+@ExtendWith(SpringExtension.class)
+@DataJpaTest
+public class AuthorRepositoryTest {
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:36]: In line with our style guide, could you please create a class display name (i.e. `AuthorRepository should`)?

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    private BookRepository bookRepository;
+    
+    @Autowired
+    private AuthorRepository authorRepository;
+    
+    @Test
+    public void testFindAllBooks() {
+      
+        assertNotNull(bookRepository);
+        assertNotNull(authorRepository);
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:48]: Could you please remove these?

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+public class AuthorRepositoryTest {
+  
+    @Autowired
+    private BookRepository bookRepository;
+    
+    @Autowired
+    private AuthorRepository authorRepository;
+    
+    @Test
+    public void testFindAllBooks() {
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:45]: As per our style guide, could you please rename this to follow our [test naming style](https://github.com/Project-Books/book-project/blob/master/STYLEGUIDE.md#using-displayname)?

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        bookRepository.save(book1);
+        
+        List<Author> result = authorRepository.findAllAuthors();
+        
+        assertTrue(result.stream().anyMatch(e -> e.equals(author1)));
+        assertTrue(result.stream().anyMatch(e -> e.equals(author2)));
+        
+        authorRepository.delete(author1);
+        authorRepository.delete(author2);
+        bookRepository.delete(book1);
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:71]: Could you please move this to a method called `tearDown` that runs after each test method?

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        book1.setPublishedBy(Publisher.CAMBRIDGE_UNIVERSITY_PRESS);
+        book1.setFormat(BookFormat.PAPERBACK);
+        
+        authorRepository.save(author1);
+        authorRepository.save(author2);
+        bookRepository.save(book1);
+        
+        List<Author> result = authorRepository.findAllAuthors();
+        
+        assertTrue(result.stream().anyMatch(e -> e.equals(author1)));
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:66]: Note that, as per our style guide, we're using AssertJ assertions, not JUnit assertions

File: src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import com.karankumar.booksapi.model.Author;
+import com.karankumar.booksapi.model.Book;
+import com.karankumar.booksapi.model.BookFormat;
+import com.karankumar.booksapi.model.BookGenre;
+import com.karankumar.booksapi.model.Publisher;
+
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java:33]: Comments from `AuthorRepositoryTest` also apply here

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        
+        Author author1 = new Author("Kevlin", "Henney");
+        Author author2 = new Author("Trisha", "Gee");
+        Book book1 = new Book("97 Things Every Java Programmer Should Know",
+            new Author[] {author1, author2});
+        book1.setGenre(BookGenre.REFERENCE);
+        book1.setYearOfPublication(2019);
+        book1.setIsbn13("9781408670545");
+        book1.setPublishedBy(Publisher.CAMBRIDGE_UNIVERSITY_PRESS);
+        book1.setFormat(BookFormat.PAPERBACK);
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java]: Could you please extract this to a new method?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for adding the tests. I noticed there's a number of style guide violations, so I would recommend reviewing our [style guide](https://github.com/Project-Books/book-project/blob/master/STYLEGUIDE.md#using-displayname)?

**Comment by chetanchilhate** [comment]
@knjk04 I have made suggested changes have look let me know if I missed anything I will fix it

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+@ExtendWith(SpringExtension.class)
+@DisplayName("AuthorRepository should")
+public class AuthorRepositoryTest {
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:39]: Could you please remove the `public` modifier here? Minor, but it keeps SonarCloud happy

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    public AuthorRepositoryTest(BookRepository bookRepository, AuthorRepository authorRepository) {
+        this.bookRepository = bookRepository;
+        this.authorRepository = authorRepository;
+        this.books = new ArrayList<>();
+        this.authors = new ArrayList<>();
+    }
+  
+    @Test
+    @DisplayName("find saved authors")
+    public void findSavedAuthors() {
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:57]: Could you please remove the `public` modifier?

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        saveBookAndAuthors();
+    
+        List<Author> result = authorRepository.findAllAuthors();
+    
+        assertThat(result.stream().anyMatch(e -> e.equals(authors.get(0))))
+            .isTrue();
+        
+        assertThat(result.stream().anyMatch(e -> e.equals(authors.get(1))))
+            .isTrue();
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java:66]: Could you please

- assert that the size is two?
- combine the current two assertions into one using `contains` on the `result`?

File: src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    public BookRepositoryTest(BookRepository bookRepository, AuthorRepository authorRepository) {
+        this.bookRepository = bookRepository;
+        this.authorRepository = authorRepository;
+        this.books = new ArrayList<>();
+        this.authors = new ArrayList<>();
+    }
+  
+    @Test
+    @DisplayName("find saved book and authors")
+    public void findSavedBookAndAuthors() {
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java:57]: Could you please also add an assert for the size of the result and combine the `.get(0)` and `.get(1)` assertions?

File: src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+@ExtendWith(SpringExtension.class)
+@DisplayName("BookRepository should")
+public class BookRepositoryTest {
```
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java:39]: Could you please also remove the public modifier here?

File: src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java (hunk 14 above)
Comment (knjk04) [inline src/test/java/com/karankumar/booksapi/repository/BookRepositoryTest.java:57]: Could you please also remove the public modifier here?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for making the changes so quickly! I've left a few more suggestions for improvements

File: src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java (hunk 13 above)
Comment (chetanchilhate) [reply src/test/java/com/karankumar/booksapi/repository/AuthorRepositoryTest.java]: added assertion to check size should be grater than or equal to 2, we could have existing values in table

**Review by knjk04 — APPROVED** [review]
Thanks for implementing the changes! I refactored the code, so let me know if you have any comments

**Pull request merged**